packages/w3up-client/src/client.js

```diff
@@ -311,4 +311,81 @@ export class Client extends Base {
       proofs: options.proofs,
     })
   }
+
+  /**
+   * Removes association of a content CID with the space. Optionally, also removes
+   * association of CAR shards with space.
+   *
+   * ⚠️ Please be aware that all shards will be deleted even if there is another upload(s) that
+   * reference same shards, which in turn could corrupt those uploads.
+   *
+   * @param {import('multiformats').UnknownLink} contentCID
+   * @param {object} [options]
+   * @param {boolean} [options.shards]
+   */
+  async remove(contentCID, options = {}) {
+    // Shortcut if there is no request to remove shards
+    if (!options.shards) {
+      // Remove association of content CID with selected space.
+      try {
+        await this.capability.upload.remove(contentCID)
+        /* c8 ignore start */
+      } catch (/** @type {any} */ err) {
+        throw new Error(
+          `remove failed for ${contentCID}: ${err.message ?? err}`
+        )
+      }
+      return
+    }
+
+    // Get shards associated with upload.
+    let upload
+    try {
+      upload = await this.capability.upload.get(contentCID)
+      /* c8 ignore start */
+    } catch (/** @type {any} */ err) {
+      throw new Error(`remove failed for ${contentCID}: ${err.message ?? err}`)
+    }
+    /* c8 ignore stop */
+
+    // Check if we have information about the shards to remove
+    if (!upload.root) {
+      throw new Error(
+        `⁂ upload not found. could not determine shards to remove.`
+      )
+    }
+    /* c8 ignore start */
+    if (!upload.shards || !upload.shards.length) {
+      return
+    }
+    /* c8 ignore stop */
+
+    // Remove shards
+    await Promise.allSettled(
+      upload.shards.map(async (shard) => {
+        try {
+          await this.capability.store.remove(shard)
+        } catch (/** @type {any} */ error) {
+          /* c8 ignore start */
+          // If not found, we can tolerate error as it may be a consecutive call for deletion where first failed
+          if (error?.cause?.name !== 'StoreItemNotFound') {
+            throw new Error(
+              `remove failed for shard ${shard.link()} from ${contentCID}: ${
+                error?.message
+              }`
+            )
+          }
+          /* c8 ignore stop */
+        }
+      })
+    )
+
+    // Remove association of content CID with selected space.
+    try {
+      await this.capability.upload.remove(contentCID)
+      /* c8 ignore start */
+    } catch (/** @type {any} */ err) {
+      throw new Error(`remove failed for ${contentCID}: ${err.message ?? err}`)
+    }
```

Comment:
Personally I'd wrap a conditional around the shard deleting bit and not repeat this try/catch twice in this same function.